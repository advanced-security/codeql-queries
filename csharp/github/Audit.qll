private import semmle.code.csharp.dataflow.DataFlow::DataFlow::PathGraph
private import semmle.code.csharp.dataflow.ExternalFlow

class AuditSinks extends DataFlow::Node {
  AuditSinks() { sinkNode(this, "audit") }
}
