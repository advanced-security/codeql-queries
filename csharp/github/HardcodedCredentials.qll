
private import csharp
private import semmle.code.csharp.frameworks.Moq
private import semmle.code.csharp.frameworks.System
private import semmle.code.csharp.dataflow.TaintTracking

/**
 * A data flow source for hard coded credentials.
 */
abstract class Source extends DataFlow::ExprNode { }

/**
 * A data flow sink for hard coded credentials.
 */
abstract class Sink extends DataFlow::ExprNode { }

/**
 * A sanitizer for hard coded credentials.
 */
abstract class Sanitizer extends DataFlow::ExprNode { }

/**
 * A string literal that is not empty, and not an expression placeholder.
 */
class NonEmptyStringLiteral extends Source {
  NonEmptyStringLiteral() {
    exists(StringLiteral literal |
      this.getExpr() = literal
      and literal.getValue().length() > 1
      and literal.fromSource()
      and not literal.getValue().matches("${%}")
    )
  }
}

/**
 * The creation of a literal byte array.
 */
class ByteArrayLiteral extends Source {
  ByteArrayLiteral() {
    this.getExpr() =
      any(ArrayCreation ac |
        ac.getArrayType().getElementType() instanceof ByteType and
        ac.hasInitializer()
      )
    and this.getExpr().fromSource()
  }
}

/**
 * The creation of a literal char array.
 */
class CharArrayLiteral extends Source {
  CharArrayLiteral() {
    this.getExpr() =
      any(ArrayCreation ac |
        ac.getArrayType().getElementType() instanceof CharType and
        ac.hasInitializer()
      )
    and this.getExpr().fromSource()
  }
}

// taint from a string literal to the constructor of a SymmetricSecurityKey
class LiteralToSecurityKeyConfig extends TaintTracking::Configuration {
  LiteralToSecurityKeyConfig() { this = "LiteralToSecurityKeyConfig" }

  override predicate isSource(DataFlow::Node source) { source instanceof Source }

  override predicate isSink(DataFlow::Node sink) {
    sink instanceof Sink
  }

  override predicate isSanitizer(DataFlow::Node node) { node instanceof Sanitizer }
}

class SymmetricSecurityKey extends Sink {
  SymmetricSecurityKey() {
    exists(ObjectCreation securityKey |
      securityKey.getAnArgument() = this.getExpr() and
      securityKey.getType().getQualifiedName() in [
        "Microsoft.IdentityModel.Tokens.SymmetricSecurityKey",
        "System.IdentityModel.Tokens.SymmetricSecurityKey"
      ]
    )
  }
}


/** 
 *  A result from a mock interface
 *  Ignore values that are ultimately returned by mocks, as they don't represent "real" credentials.
*/
class MockSanitizer extends Sanitizer {
  MockSanitizer() {
    exists(ReturnedByMockObject mock|
      mock.getAMemberInitializationValue() = this.asExpr()
      or mock.getAnArgument() = this.asExpr()
    )
  }
}

/** 
 *  A result from a mock interface
 *  The function is not itself marked as a mock, but all uses of it are in mocks.
*/
class MockSanitizer2 extends Sanitizer {
  MockSanitizer2() {
    exists(Method method, ReturnedByMockObject mock |
      exists(Call call |
        call = method.getACall() and method.getAChild*() = this.asExpr() and
        (
          mock.getAMemberInitializationValue().getAChild*() = call
          or mock.getAnArgument().getAChild*() = call
        )
      )
      and not exists(Call call |
        call = method.getACall() and method.getAChild*() = this.asExpr() and
        not (
          mock.getAMemberInitializationValue().getAChild*() = call
          or mock.getAnArgument().getAChild*() = call
        )
        )
    )
  }
}

/**
 * A format argument to `Format`, that is considered not to be a source of hardcoded secret data.
 */
class StringFormatSanitizer extends Sanitizer {
  StringFormatSanitizer() {
    this.getExpr() =
      any(SystemStringClass s).getFormatMethod().getACall().getArgumentForName("format")
  }
}

/**
 * A replacement argument to `Replace`, that is considered not to be a source of hardcoded secret
 * data.
 */
class StringReplaceSanitizer extends Sanitizer {
  StringReplaceSanitizer() {
    exists(SystemStringClass s, Call c | c = s.getReplaceMethod().getACall() |
      this.getExpr() = c.getArgumentForName("newValue") or
      this.getExpr() = c.getArgumentForName("newChar")
    )
  }
}

/**
 * A call to a `ToString()` method, which is considered not to return hard-coded constants.
 */
class ToStringSanitizer extends Sanitizer {
  ToStringSanitizer() { this.getExpr() = any(Call c | c.getTarget().hasName("ToString")) }
}

/**
 * A call to a configuration section or access to configuration data, using the string as a key.
 */
class ConfigurationSanitizer extends Sanitizer {
  ConfigurationSanitizer() {
    exists(Access configuration, MethodCall call |
      configuration.getType().getQualifiedName() in [
        "Microsoft.Extensions.Configuration.IConfiguration", "Microsoft.Extensions.Configuration.ConfigurationManager"
      ]
      and call.getQualifier() = configuration
      and call.getAnArgument() = this.getExpr() 
    )
  }
}

/**
 * A call to a file operation, opening or reading from a named file.
 */
class FileSanitizer extends Sanitizer {
  FileSanitizer() {
    exists(Call c | c.getTarget().hasQualifiedName("System.IO.File", [
      "ReadAllBytes", "ReadAllText", "Open", "OpenText", "OpenRead", "OpenHandle", "ReadAllTextAsync", "ReadAllBytesAsync", "ReadAllLines", "ReadAllLinesAsync", "ReadLines", "ReadLinesAsync", "OpenTextAsync"
    ]) and
      c.getAnArgument() = this.getExpr()
    )
  }
}

/**
 * An expr within a method of a Class with a TestClass or Fact attribute, or a name that suggests testing
 */
class TestClassSanitizer extends Sanitizer {
  TestClassSanitizer() {
    exists(Class c |
      (
        c.getAnAttribute().getType().getName() = ["TestClassAttribute", "FactAttribute"] or
        c.getName().matches(["Test%", "%Test", "%Tests", "Mock%", "%Mocks", "%Mock", "Fake%"])
      ) and
      (
        this.getExpr() = c.getAMethod().getAChild*() or
        this.getExpr() = c.getAField().getAChild*() or
        this.getExpr() = c.getAProperty().getGetter().getBody().getAChild*()
      )
    )
  }
}

/**
 * An expr within a method of a Class with inside a Namespace that suggests testing
 */
class TestNamespaceSanitizer extends Sanitizer {
  TestNamespaceSanitizer() {
    exists(Namespace n | n.getName().matches(["Test%", "%Test", "%Tests", "Mock%", "%Mocks", "%Mock", "Fake%"]) and
      (
        this.getExpr() = n.getAClass().getAMethod().getAChild*() or
        this.getExpr() = n.getAClass().getAField().getAChild*() or
        this.getExpr() = n.getAClass().getAProperty().getGetter().getBody().getAChild*()
      )
    )
  }
}

/**
 * A preprocessor directive for DEBUG, so this isn't a real secret used in a Release build.
 */
class DebugSanitizer extends Sanitizer {
  DebugSanitizer() {
    exists(IfDirective if_d, EndifDirective endif_d, Location if_loc, Location endif_loc, Location loc |
      loc = this.getLocation() and
      if_d.getCondition().toString() = "DEBUG" and if_d.getEndifDirective() = endif_d
      and if_d.getLocation() = if_loc and endif_d.getLocation() = endif_loc
      and loc.getStartLine() > if_loc.getEndLine()
      and loc.getEndLine() < endif_loc.getStartLine()
      and loc.getFile() = if_loc.getFile()
    )
  }
}