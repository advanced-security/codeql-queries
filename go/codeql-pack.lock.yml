---
lockVersion: 1.0.0
dependencies:
  codeql/go-all:
    version: 0.5.2
  codeql/tutorial:
    version: 0.0.10
  codeql/util:
    version: 0.0.10
compiled: false
