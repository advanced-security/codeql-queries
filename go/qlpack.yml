name: github-queries-go
version: 0.1.0
dependencies:
  codeql/go-queries: "*"