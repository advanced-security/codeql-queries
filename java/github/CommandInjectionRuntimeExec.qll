import java
import semmle.code.java.frameworks.javaee.ejb.EJBRestrictions
import semmle.code.java.dataflow.DataFlow
private import semmle.code.java.dataflow.TaintTracking
import semmle.code.java.dataflow.FlowSources


// a static string of an unsafe executable tainting arg 0 of Runtime.exec()
class ExecTaintConfiguration extends TaintTracking::Configuration {
    ExecTaintConfiguration() { this = "ExecTaintConfiguration" }

    override
    predicate
    isSource(DataFlow::Node source) {
        source.asExpr() instanceof StringLiteral
        and source.asExpr().(StringLiteral).getValue() instanceof UnSafeExecutable
    }

    override
    predicate
    isSink(DataFlow::Node sink) {
        exists(RuntimeExecMethod method, MethodAccess call |
            call.getMethod() = method
            and sink.asExpr() = call.getArgument(0)
            and sink.asExpr().getType() instanceof Array
        )
    }

    override
    predicate
    isSanitizer(DataFlow::Node node) {
        node.asExpr().getFile().isSourceFile() and
        (
            node instanceof AssignToNonZeroIndex
            or node instanceof ArrayInitAtNonZeroIndex
            or node instanceof StreamConcatAtNonZeroIndex
            or node.getType() instanceof PrimitiveType
            or node.getType() instanceof BoxedType
        )
    }
}

abstract class Source extends DataFlow::Node {
    Source() {
        this = this
    }
}


// taint flow from user data to args of the command
class ExecTaintConfiguration2 extends TaintTracking::Configuration {
    ExecTaintConfiguration2() { this = "ExecTaintConfiguration2" }

    override
    predicate
    isSource(DataFlow::Node source) {
        source instanceof Source
    }

    override
    predicate
    isSink(DataFlow::Node sink) {
        exists(RuntimeExecMethod method, MethodAccess call, int index |
            call.getMethod() = method
            and sink.asExpr() = call.getArgument(index)
            and sink.asExpr().getType() instanceof Array
        )
    }

    override
    predicate
    isSanitizer(DataFlow::Node node) {
        node.asExpr().getFile().isSourceFile() and
        (
            node.getType() instanceof PrimitiveType
            or node.getType() instanceof BoxedType
        )
    }
}


// array[3] = node
class AssignToNonZeroIndex extends DataFlow::Node {
    AssignExpr assign;
    ArrayAccess access;

    AssignToNonZeroIndex() {
        assign.getDest() = access
        and access.getIndexExpr().(IntegerLiteral).getValue() != "0"
        and assign.getSource() = this.asExpr()
    }
}


// String[] array = {"a", "b, "c"};
class ArrayInitAtNonZeroIndex extends DataFlow::Node {
    ArrayInit init;
    int index;

    ArrayInitAtNonZeroIndex() {
        init.getInit(index) = this.asExpr()
        and index != 0
    }
}

// Stream.concat(Arrays.stream(array_1), Arrays.stream(array_2))
class StreamConcatAtNonZeroIndex extends DataFlow::Node {
    MethodAccess call;
    int index;

    StreamConcatAtNonZeroIndex() {
        call.getMethod().getQualifiedName() = "java.util.stream.Stream.concat"
        and call.getArgument(index) = this.asExpr()
        and index != 0
    }
}


// allow list of executables that execute their arguments
// TODO: extend with data extensions
class UnSafeExecutable extends string {
    bindingset[this]
    UnSafeExecutable() {
        this.regexpMatch("^(|.*/)([a-z]*sh|javac?|python[23]?|perl|[Pp]ower[Ss]hell|php|node|deno|bun|ruby|osascript|cmd|Rscript|groovy)(\\.exe)?$")
        and not this.matches("netsh.exe")
    }
}
