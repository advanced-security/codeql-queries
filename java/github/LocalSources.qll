import java
import semmle.code.java.dataflow.FlowSources

class FileReadAccess extends LocalUserInput {
  FileReadAccess() {
    exists(MethodAccess ma |
      // https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getResourceAsStream-java.lang.String-
      (
        ma.getMethod().hasQualifiedName("java.lang", "Class", "getResourceAsStream") or
        ma.getMethod().hasQualifiedName("java.lang", "ClassLoader", "getResourceAsStream") or
        // Finds any method that has the name "getResourceAsStream"
        ma.getMethod().hasQualifiedName(_, _, "getResourceAsStream")
      ) and
      this.asExpr() = ma.getArgument(0)
    )
  }
}
