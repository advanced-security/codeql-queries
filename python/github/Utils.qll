import python
private import semmle.python.ApiGraphs
private import semmle.python.Concepts
private import semmle.python.dataflow.new.DataFlow
private import semmle.python.dataflow.new.internal.TaintTrackingPrivate

// List of all the format strings
// - python/ql/lib/semmle/python/dataflow/new/internal/TaintTrackingPrivate.qll
class DynamicStrings extends DataFlow::Node {
  DynamicStrings() {
    (
      // s = f"WHERE name = '{input}'"
      exists(Fstring fmtstr | this.asExpr() = fmtstr)
      or
      // "SELECT * FROM users WHERE username = '{}'".format(username)
      exists(CallNode format, string methods, ControlFlowNode object |
        object = format.getFunction().(AttrNode).getObject(methods)
      |
        methods = "format" and
        this.asExpr() = format.getNode()
      )
      or
      exists(BinaryExpr expr |
        (
          // q = "WHERE name = %s" % username
          expr.getOp() instanceof Mod or
          // q = "WHERE name = " + username
          expr.getOp() instanceof Add
        )
        and
        expr.getLeft().getParent() = this.asExpr()
      )
    ) and
    this.getScope().inSource()
  }
}
