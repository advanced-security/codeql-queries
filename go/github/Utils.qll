import go
import semmle.go.frameworks.stdlib.Fmt

class DynamicStrings extends DataFlow::Node {
    DynamicStrings() {
        // fmt format string
        exists(Fmt::Sprinter formatter |
            this = formatter.getACall()
        )
        or
        // binary expression
        exists(BinaryExpr expr |
            this.asExpr() = expr.getLeftOperand() and
            expr.getOperator() = "+"
        )
    }
}