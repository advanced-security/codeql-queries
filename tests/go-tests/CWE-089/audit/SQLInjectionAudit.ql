import go
import github.Utils

query predicate dynamicStrings(DataFlow::ExprNode node) { node instanceof DynamicStrings }
