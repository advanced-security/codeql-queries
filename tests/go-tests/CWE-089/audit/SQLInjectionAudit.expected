| main.go:27:21:27:56 | "SELECT * FROM users WHERE name = '" |
| main.go:27:21:27:64 | ...+... |
| main.go:31:22:31:80 | call to Sprintf |
